### PR TITLE
Add redis

### PR DESCRIPTION
**Pull request by MosesNgan**

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -1,9 +1,41 @@
...
+      return res.send(cachedData);
+    }
+
+    // If data is not found in cache, continue to the next middleware/route
+    return next();
+  } catch (error) {
+    console.error('Error accessing Redis cache:', error);
+    return res.status(500).send('Internal Server Error');
+  }
+};
```
Comment (MosesNgan) [inline src/app.ts]: Add cache middleware

File: src/services/cache.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  return redisClient;
+}
+
+async function getRedisClient(): Promise<RedisClientType> {
+  if (!isReady) {
+    redisClient = await createRedisClient();
+  }
+
+  return redisClient;
+}
```
Comment (MosesNgan) [inline src/services/cache.ts:36]: refactor: connect redis ONCE only for upcoming get and set calls

File: docker-compose.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+      context: .
+      dockerfile: Dockerfile
+    environment:
+      NODE_ENV: development
+      REDIS_URL: redis://redis:6379
+    ports:
+      - 3000:3000
+    volumes:
+      - .:/app
+      - /app/node_modules
```
Comment (MosesNgan) [inline docker-compose.yaml:14]: mounting the source code

File: src/app.ts (hunk 1 above)
Comment (domwongpaloit) [reply src/app.ts]: usually we don't use 500, it is for real disaster happens.

**Pull request merged**